sql: selectNode.Next returns false if renderRow errors

dt opened the pull request:
Both Next() and Values() may cause plan.Err() to return non-nil.
It is assumed that on error the result cannot be used, but
failing to check that meant executor was hitting an NPE.

Fixes #6601.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6608)

<!-- Reviewable:end -->

RaduBerinde commented:
Hi David,

The convention is:
 - if `Next()` hits an error, it sets an error **and** returns `false`. Seems like that is not happening in this case. We should find where and fix it.
 - if `Next()` returns `true`, `Values()` works. Calling `Values()` would never set an error.
These new checks (if anything) should be panics.

<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> #### sql: Stop reading plan on err
> 
> Both Next() and Values() may cause plan.Err() to return non-nil.
> 
> It is assumed that on error the result cannot be used, but
> 
> failing to check that meant executor was hitting an NPE.
> 
> 
> 
> Fixes #6601.
> 
> 
> </details>


---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-:-KHPvhI4gwICBIQguajR:1428372131)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
(see https://github.com/cockroachdb/cockroach/blob/master/sql/plan.go#L432)

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Hi David,
> 
> The convention is:
>  - if `Next()` hits an error, it sets an error **and** returns `false`. Seems like that is not happening in this case. We should find where and fix it.
>  - if `Next()` returns `true`, `Values()` works. Calling `Values()` would never set an error.
> These new checks (if anything) should be panics.</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-:-KHPwvoO4LqW7gk3yNei:604620544)*
<!-- Sent from Reviewable.io -->

dt commented:
Ah, I think the issue here is that the Values() call *is* encountering an
error. So either the work it was doing needs to move to Next or we have to
check Err().

On Tue, May 10, 2016 at 9:56 AM RaduBerinde <notifications@github.com>
wrote:

> (see https://github.com/cockroachdb/cockroach/blob/master/sql/plan.go#L432
> )
>
> *Previously, RaduBerinde wrote…*
>
> Hi David,
>
> The convention is:
>
>    - if Next() hits an error, it sets an error *and* returns false. Seems
>    like that is not happening in this case. We should find where and fix it.
>    - if Next() returns true, Values() works. Calling Values() would never
>    set an error. These new checks (if anything) should be panics.
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-:-KHPwvoO4LqW7gk3yNei:604620544>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6608#issuecomment-218164649>
>

RaduBerinde commented:
Yeah, anything that can fail should be in `Next()`. Usually `Values()` is just a getter for something that was populated during `Next()`.

If we would have to change this convention, there is a lot of code that needs to be adjusted - anywhere we loop through a plan (aggregation, sorting, select, etc). Pretty much every call site of `Values()`.

<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> Ah, I think the issue here is that the Values() call *is* encountering an error. So either the work it was doing needs to move to Next or we have to check Err().
> 
> </details>


---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-:-KHPy1RIsWYRmybLXDs9:-1438747725)*
<!-- Sent from Reviewable.io -->

dt committed:
sql: selectNode.Next returns false if renderRow errors

Fixes #6601.

dt commented:
Ah, turns out it _was_ in `Next` but `Next` was returning true while p.err was non-nil.

That said, if the convention is that `Next` has to do all the work anyway, then I think it'd make sense just to let it return the result, rather than handling more complex states. OTOH, if we want to be able to more cheaply call `Next` when we don't actually need the values (eg counting or skipping), then whatever *does* do that work, e.g. `Values()`, would seem to want to be able to signal an error.

RaduBerinde commented:
There is one caveat there - in the special 'explain debug' mode, `Next()` doesn't guarantee that `Values()` will work - it guarantees that `DebugValues()` works, and those results indicate if `Values()` would also work for that row. Pretty annoying. We would either have to return both kinds of values (which "pollutes" the "normal" path with this extra stuff), or maybe introduce a `nextDebug()` that's only used for that mode.

RaduBerinde commented:
LGTM, good catch!

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> There is one caveat there - in the special 'explain debug' mode, `Next()` doesn't guarantee that `Values()` will work - it guarantees that `DebugValues()` works, and those results indicate if `Values()` would also work for that row. Pretty annoying. We would either have to return both kinds of values (which "pollutes" the "normal" path with this extra stuff), or maybe introduce a `nextDebug()` that's only used for that mode.</details>


---

Reviewed 1 of 2 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/select.go, line 135 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQ4B4m4zJcO22q-9pQ:-KHQ4B4m4zJcO22q-9pR:887143682) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d56d7db76eeeef0142c48d1a57359c3ac8bf0bf/sql/select.go#L135)):*
> ```Go
> 
> 		if passesFilter {
> 			s.renderRow()
> ```

maybe we should make `renderRow` return the error and set `s.err` here, for consistency with the `RunFilter `case above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-:-KHQ4Hsl71CVYJaXYh2X:-1384602500)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/select.go, line 135 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQ4B4m4zJcO22q-9pQ:-KHQ5CF2CKnXGAe6O3U5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d56d7db76eeeef0142c48d1a57359c3ac8bf0bf/sql/select.go#L135)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> maybe we should make `renderRow` return the error and set `s.err` here, for consistency with the `RunFilter `case above.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
nice catch!

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> LGTM, good catch!</details>


---

Reviewed 1 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-:-KHQ5Z0N40u0aIP2PQyI:1412495856)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> :lgtm:
> nice catch!</details>


---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/testdata/builtin_function, line 116 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQ5_xAzLAd0dTHQqsR:-KHQ5_xAzLAd0dTHQqsS:-884752015) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8fd6aa7b3eed2c416042dc39d07837b0cb4a43/sql/testdata/builtin_function#L116)):*
> ```
> 
> # See #6601
> query error invalid utf8: .*
> ```

i dont think you need the trailing `.*`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-:-KHQ5ibPH9iBML4HbicR:1417088721)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/select.go, line 623 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQ6U1NN2QmoRxMibcA:-KHQ6U1NN2QmoRxMibcB:43967903) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8fd6aa7b3eed2c416042dc39d07837b0cb4a43/sql/select.go#L623)):*
> ```Go
> 		var err error
> 		s.row[i], err = e.Eval(s.planner.evalCtx)
> 		if s.err != nil {
> ```

Are we setting `s.err` anymore? Shouldnt this be `if err != nil {`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/select.go, line 623 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQ6U1NN2QmoRxMibcA:-KHQDB_fPyZUf5xeD5NO:161152953) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8fd6aa7b3eed2c416042dc39d07837b0cb4a43/sql/select.go#L623)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Are we setting `s.err` anymore? Shouldnt this be `if err != nil {`</details>
Aha, yep, fixed. Also, turns out that when I switched the test case to use a byte literal rather than `uuid_v4` it stopped failing so this wasn't caught. reverted that too.

---

*[sql/testdata/builtin_function, line 116 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQ5_xAzLAd0dTHQqsR:-KHQDPdMKkhyapFINwHp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8fd6aa7b3eed2c416042dc39d07837b0cb4a43/sql/testdata/builtin_function#L116)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i dont think you need the trailing `.*`</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/testdata/builtin_function, line 120 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQE9fbs2s9dsEmVq7T:-KHQE9fcBEmAVYPl0udf:1426426337) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8fd6aa7b3eed2c416042dc39d07837b0cb4a43/sql/testdata/builtin_function#L120)):*
> ```
> 
> query error invalid utf8: .*
> explain(debug) select b'\x61\xa1'::string where false;
> ```

you don't want to keep the debug test?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/testdata/builtin_function, line 117 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQJrtrw7mZ-hZrtdO2:-KHQJrtsQ9YikMVb4cf3:-1899262000) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2e269990fa3a52a127418784844dd0f8ce2f2da/sql/testdata/builtin_function#L117)):*
> ```
> # See #6601
> query error invalid utf8
> select cast(uuid_v4() as string)
> ```

How unlikely is `uuid_v4` to return a valid UTF8 string? If it's completely random, there's `(1/2)^len` chance that all bytes are ASCII.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/testdata/builtin_function, line 120 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQE9fbs2s9dsEmVq7T:-KHQLsNVXcxbrHHBBuvl:1009663016) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8fd6aa7b3eed2c416042dc39d07837b0cb4a43/sql/testdata/builtin_function#L120)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you don't want to keep the debug test?</details>
When I looked closer, it wasn't actually testing anything.

---

*[sql/testdata/builtin_function, line 117 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608#-KHQJrtrw7mZ-hZrtdO2:-KHQLjHH5Zc_PC1I3VE_:650128570) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2e269990fa3a52a127418784844dd0f8ce2f2da/sql/testdata/builtin_function#L117)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> How unlikely is `uuid_v4` to return a valid UTF8 string? If it's completely random, there's `(1/2)^len` chance that all bytes are ASCII.</details>
yeah, at the very least the variant byte is and'ed with 0x80 so it should be always non-UTF8, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6608)*
<!-- Sent from Reviewable.io -->
